pkg/threescale/backend/manager.go

```diff
@@ -149,6 +161,45 @@ func (b *Backend) Report(request threescale.Request) (*threescale.ReportResult,
 	return &threescale.ReportResult{Accepted: true}, nil
 }
 
+// Flush the cached entries and report existing state to backend
+func (b *Backend) Flush() {
+	// create a hash of cached applications under a service ID
+	// this may not be required down in future and we can simplify but
+	// this gives us the flexibility to do batched reporting
+	groupedKeys := b.groupCachedItems()
+	for _, service := range groupedKeys {
+
+		for i, cachedApp := range service {
+			app, ok := b.cache.Get(cachedApp.cacheKey)
+			if !ok {
+				continue
+			}
+			// for each application for this service
+			// take a read lock and take a snapshot of the current state
+			app.RLock()
+			cachedApp.snapshot = app.deepCopy()
```

Comment:
Once we create the snapshot we should be able to unlock the original app and use instead the snapshot in the code below. After we report succesfully we could lock the app for writing and then deduce the reported values from the now potentially updated/incremented ones.